File: python/append_digit_primes.py
import math

from tree import Tree
import util

def next(x):
    new_xs = []
    for i in range(10):
        if util.is_prime(10 * x + i):
            new_xs.append(10 * x + i)
    return new_xs


def main():
    tree = Tree(0, [])
    for i in range(20):
        print(tree.longest_path())
        tree.step(next)
    print(tree.longest_path())


if __name__ == '__main__':
    main()

File: cpp/append_digit_primes/append_digit_primes.cpp
#include "../common/util.h"
#include "../common/tree.h"

#include <cstdlib>

auto next(ull x) -> std::vector<ull> {
    std::vector<ull> new_xs;
    ull temp = x * 10;
    for (int i = 0; i < 10; i++) {
        auto temp2 = temp + i;
        if (is_prime(temp2)) {
            new_xs.push_back(temp2);
        }
    }
    return new_xs;
}

auto main(int argc, char *argv[]) -> int {
    if (argc != 2) {
        std::cout << "Usage: append_digit_primes <search_depth>\n";
        return -1;
    }
    Tree tree(0, {});
    for (int i = 0; i < std::atoi(argv[1]); i++) {
        tree.step(next);
        print_vec(tree.longest_path());
    }
}
